In [1]:
# Styling notebook
from IPython.core.display import HTML
def css_styling():
    styles = open("./styles/style.css", "r").read()
    s = '<style>\n %s \n</style>' % styles
    return HTML(s)
css_styling()

<p class="title align-center persian">
  پروژه ی یک درس هوش مصنوعی
</p>
<p class="persian align-right">
  در این پروژه، خواسته شده است که به روش های جست و جوی آگاهانه و نا آگاهانه، برای یک بازی راه حلی کامل و بهینه یافته شود. در این بازی گندالف شخصیتی است که باید در زمین ابتدا هر کدام از یاران حلقه را به مقصد مشخص شده یشان برساند و سپس خود به گاندور برود. این در حالی است که در عین حال نباید در طول مسیر هیچ ارکی را بیدار کند. ارک ها در دو صورت بیدار میشوند: یک مواجه شدن با گندالف در خانه ی خود و دو اینکه گندالف برای مدتی طولانی تر از آستانه ی آن ارک در محدوده ی تحت کنترلش باشد.
</p>

<p class="persian align-right">
  برای ارائه ی راه حل ها، مسئله به صورت زیر تعریف می شود:
</p>

<ul class="persian align-right">
  <li>
    استیت ها: در هر استیت موقعیت گندالف، یاران حلقه و ارکی که گندالف در محدوده ی تحت کنترلش است ( می تواند در محدوده ی هیچ ارکی نباشد )، ثابت است.
  </li>
  <li>
    حرکت ها: حرکت ها به چهار سمت راست، چپ، بالا و پایین خلاصه می شود.
  </li>
  <li>
    استیت اولیه: استیتی است که گندالف در نقطه ی شروعش است، هیچ کدام از یاران حلقه نه همراه گندالف اند و نه در مقصدشان و  گندالف در محدوده ی هیچ ارکی نیست.
  </li>
  <li>
    استیت نهایی: استیتی است که گندالف در گاندور است و همه ی یاران به مقصدشان رسیده اند.
  </li>
</ul>

<p class="persian align-right">
  برای ارائه ی این پروژه سه پکیج utils، game و solution طراحی شده اند.
</p>

<p class="persian align-right">
  در پکیج utils کلاس های پایه ی برنامه قرار گرفته اند، در پکیج game اطلاعات مربوط به بازی قرار دارند و در پکیج solution توابع مربوط به هر یک از این الگوریتم های جست و جو پیاده سازی شده است.
</p>
<p class="persian align-right">
  به طور دقیق تر در پکیج game دو کلاس مهم را نگه می داریم: یک اطلاعات مربوط به برد که در طول بازی ثابت است و دو اطلاعات مربوط به استیت های بازی که شامل چگونگی ذخیره کردن و نحوه ی ارتباط استیت ها از طریق حرکت ها است.
</p>

In [2]:
from solution.utils.formulateTestCase import formulateTestCase
from solution.bfs.bfs import Bfs
from solution.ids.ids import Ids
from solution.aStar.aStar import AStar
from solution.weightedAStar.weightedAStar import WeightedAStar
from time import time

testCount = 4

<p class="persian align-right">
  در ابتدا هر کدام از تست کیس ها خوانده شده و به صورت یک گیم استیت اولیه و یک برد در یک لیست ذخیره می شود.
</p>

In [3]:
boards = list()
initGameStates = list()
for i in range(testCount):
  board, initGameState = formulateTestCase('../test/test_0' + str(i) + '.txt')
  boards.append(board)
  initGameStates.append(initGameState)

<p class="persian align-right">
  سپس برای هر کدام از الگوریتم ها، همه ی تست کیس ها تست شده و به ترتیب اندازه ی جواب، مسیر جواب، زمان اجرا و تعداد استیت های دیده شده به نمایش در می آید.
</p>

In [4]:
for i in range(testCount):
  start = time()
  goalNode = Bfs.solve(boards[i], initGameStates[i])
  end = time()

  print('solution length:', goalNode.solutionLength)
  print(goalNode.moves)
  print('time:', end - start)
  print('visitedStateCount:', Bfs.visitedStateCount)
  print('* ' * 20)

solution length: 48
RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDRUDRUDRUDRUDRRRRRR
time: 0.8829529285430908
visitedStateCount: 24782
* * * * * * * * * * * * * * * * * * * * 
solution length: 52
RRRRDDDDDDDDLLLRRRUUURURUDLDLDDDLRRRRRUUUUDDDDLLLRRR
time: 0.27100205421447754
visitedStateCount: 7036
* * * * * * * * * * * * * * * * * * * * 
solution length: 34
RRRRRRRDDDDLLLLDLLLRDDRUURURRRDRDD
time: 0.06599974632263184
visitedStateCount: 1678
* * * * * * * * * * * * * * * * * * * * 
solution length: 66
RRDDDDDDDDLLRDRRRRRRRRUUUUUUUUULLDDDDDLLDDDDLLLLRRRRUUUUURRRRDDDDD
time: 0.705024003982544
visitedStateCount: 18973
* * * * * * * * * * * * * * * * * * * * 


In [14]:
for i in range(testCount):
  if i == 3:
    # continue
    pass
  start = time()
  goalNode = Ids.solve(boards[i], initGameStates[i])
  end = time()
  print('solution length:', goalNode.solutionLength)
  print(goalNode.moves)
  print('time:', end - start)
  print('visitedStateCount:', Ids.visitedStateCount)
  print('* ' * 20)

solution length: 48
RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDURDURDURDURDRRRRRR
time: 5.838996410369873
visitedStateCount: 203739
* * * * * * * * * * * * * * * * * * * * 
solution length: 52
DRRDRRLLDDDDDLDURUUUURRRRLLDDDLDDUURUURRRRDDDLDLLRRR
time: 21.4470112323761
visitedStateCount: 527731
* * * * * * * * * * * * * * * * * * * * 
solution length: 34
RRRRRRRLLLLDDDDLDLLDDRUURURRRDRDDR
time: 0.6689865589141846
visitedStateCount: 37524
* * * * * * * * * * * * * * * * * * * * 
solution length: 66
DDRDRDDLDLDDDRURUUUUUUUURRRRRRRLLLLLLLDDDDDDDDLDURUUUURRRRRRRDDDDD
time: 105.68904781341553
visitedStateCount: 1708030
* * * * * * * * * * * * * * * * * * * * 


<p class="persian align-right">
  برای الگوریتم های A Star و Weighted A Star در هر استیت از فاصله ی منهتن موقعیت گندالف در آن استیت تا نقطه ی هدف به عنوان تابع heuristic استفاده می شود.
</p>

In [6]:
for i in range(testCount):
  start = time()
  goalNode = AStar.solve(boards[i], initGameStates[i])
  end = time()
  print('solution length:', goalNode.solutionLength)
  print(goalNode.moves)
  print('time:', end - start)
  print('visitedStateCount:', AStar.visitedStateCount)
  print('* ' * 20)

solution length: 48
RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDRUDRUDRUDRUDRRRRRR
time: 0.25899815559387207
visitedStateCount: 8130
* * * * * * * * * * * * * * * * * * * * 
solution length: 52
RRRRDDDDDDLLDLDURURRURURULDDLDDDLRRRRURUUUDDDLDLLRRR
time: 0.14304280281066895
visitedStateCount: 6067
* * * * * * * * * * * * * * * * * * * * 
solution length: 36
RRRRDRRRUDDDLLLLDDLLLDDRRUURURRDRRDD
time: 0.02800130844116211
visitedStateCount: 1288
* * * * * * * * * * * * * * * * * * * * 
solution length: 66
RDDDRDDDDDLLRDRRRRRRRRUUUUUUUUULLDDDDLLDDDLLDLLDRRRUUURUURRRRDDDDD
time: 0.6359939575195312
visitedStateCount: 16187
* * * * * * * * * * * * * * * * * * * * 


In [13]:
for i in range(testCount):
  start = time()
  goalNode = WeightedAStar.solve(boards[i], initGameStates[i], 2)
  end = time()
  print('solution length:', goalNode.solutionLength)
  print(goalNode.moves)
  print('time:', end - start)
  print('visitedStateCount:', WeightedAStar.visitedStateCount)
  print('* '  * 20)

solution length: 48
RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDRUDRUDRUDRUDRRRRRR
time: 0.20892548561096191
visitedStateCount: 7506
* * * * * * * * * * * * * * * * * * * * 
solution length: 54
RRRRDDRDDDLDDDLLLRRRUUURUURDLDLDDLDRRRRRUUUUDDDDLLLRRR
time: 0.07800579071044922
visitedStateCount: 5015
* * * * * * * * * * * * * * * * * * * * 
solution length: 36
RRRRRRRDDDDLLLLDLLLRRDDLRUURURRRDRDD
time: 0.012005329132080078
visitedStateCount: 1084
* * * * * * * * * * * * * * * * * * * * 
solution length: 70
RRDDDRDRDDDLDLLLRDRRRRRRRRUUUUUUUUULLDDLDDDLDDDLLDLLRRRUUURUURRRRDDDDD
time: 0.22599196434020996
visitedStateCount: 15791
* * * * * * * * * * * * * * * * * * * * 


In [12]:
for i in range(testCount):
  start = time()
  goalNode = WeightedAStar.solve(boards[i], initGameStates[i], 3)
  end = time()
  print('solution length:', goalNode.solutionLength)
  print(goalNode.moves)
  print('time:', end - start)
  print('visitedStateCount:', WeightedAStar.visitedStateCount)
  print('* '  * 20)

solution length: 48
RRRRRRRRRRRRRRRRRRRRRRRRRRRRRDRUDRUDRUDRUDRRRRRR
time: 0.05102896690368652
visitedStateCount: 3694
* * * * * * * * * * * * * * * * * * * * 
solution length: 62
RRRRRDDLRRDDLDLDLDLLDRRRUUURUURRRDDDDDLLLLLRRRRRUUUUDDDDLLLRRR
time: 0.1229712963104248
visitedStateCount: 4598
* * * * * * * * * * * * * * * * * * * * 
solution length: 36
RRRRRRRDDDDLLLLDLLLRRDDLRUURURRRDRDD
time: 0.010999441146850586
visitedStateCount: 942
* * * * * * * * * * * * * * * * * * * * 
solution length: 76
RRDDDRDRRDDDDLLLLLRDRRRRRRRRUUUUUUUUULLRRDDDDDDDDDLLLLLLLLRRRRUUUUURRRRDDDDD
time: 0.21300220489501953
visitedStateCount: 15485
* * * * * * * * * * * * * * * * * * * * 


<p class="persian align-right">
  در اینجا با استفاده از الگوریتم های مختلف، اندازه ی جواب، تعداد استیت های دیده شده و زمان اجرا برای هر تست کیس نمایش داده شده است. 
</p>

<table class="persian align-center">
  <tr>
    <th>الگوریتم</th>
    <th>زمان اجرا</th>
    <th>اندازه ی جواب</th>
    <th>تعداد استیت ها</th>
  </tr>
  <tr>
    <td>BFS</td>
    <td>397</td>
    <td>48</td>
    <td>24782</td>
  </tr>
  <tr>
    <td>IDS</td>
    <td>5519</td>
    <td>48</td>
    <td>203739</td>
  </tr>
  <tr>
    <td>A Star</td>
    <td>167</td>
    <td>48</td>
    <td>8130</td>
  </tr>
  <tr>
    <td>Weighted A Star alpha=2</td>
    <td>161</td>
    <td>48</td>
    <td>7506</td>
  </tr>
  <tr>
    <td>Weighted A Star alpha=3</td>
    <td>62</td>
    <td>48</td>
    <td>3694</td>
  </tr>
</table>

<table class="persian align-center">
  <tr>
    <th>الگوریتم</th>
    <th>زمان اجرا</th>
    <th>اندازه ی جواب</th>
    <th>تعداد استیت ها</th>
  </tr>
  <tr>
    <td>BFS</td>
    <td>75</td>
    <td>52</td>
    <td>7036</td>
  </tr>
  <tr>
    <td>IDS</td>
    <td>18363</td>
    <td>52</td>
    <td>527731</td>
  </tr>
  <tr>
    <td>A Star</td>
    <td>72</td>
    <td>52</td>
    <td>6067</td>
  </tr>
  <tr>
    <td>Weighted A Star alpha=2</td>
    <td>72</td>
    <td>54</td>
    <td>5015</td>
  </tr>
  <tr>
    <td>Weighted A Star alpha=3</td>
    <td>125</td>
    <td>62</td>
    <td>4598</td>
  </tr>
</table>

<table class="persian align-center">
  <tr>
    <th>الگوریتم</th>
    <th>زمان اجرا</th>
    <th>اندازه ی جواب</th>
    <th>تعداد استیت ها</th>
  </tr>
  <tr>
    <td>BFS</td>
    <td>15</td>
    <td>34</td>
    <td>1678</td>
  </tr>
  <tr>
    <td>IDS</td>
    <td>920</td>
    <td>34</td>
    <td>37524</td>
  </tr>
  <tr>
    <td>A Star</td>
    <td>12</td>
    <td>34</td>
    <td>1288</td>
  </tr>
  <tr>
    <td>Weighted A Star alpha=2</td>
    <td>11</td>
    <td>36</td>
    <td>1084</td>
  </tr>
  <tr>
    <td>Weighted A Star alpha=3</td>
    <td>10</td>
    <td>36</td>
    <td>942</td>
  </tr>
</table>

<table class="persian align-center">
  <tr>
    <th>الگوریتم</th>
    <th>زمان اجرا</th>
    <th>اندازه ی جواب</th>
    <th>تعداد استیت ها</th>
  </tr>
  <tr>
    <td>BFS</td>
    <td>334</td>
    <td>66</td>
    <td>18973</td>
  </tr>
  <tr>
    <td>IDS</td>
    <td>100223</td>
    <td>66</td>
    <td>1708030</td>
  </tr>
  <tr>
    <td>A Star</td>
    <td>265</td>
    <td>66</td>
    <td>16187</td>
  </tr>
  <tr>
    <td>Weighted A Star alpha=2</td>
    <td>232</td>
    <td>70</td>
    <td>15791</td>
  </tr>
  <tr>
    <td>Weighted A Star alpha=3</td>
    <td>236</td>
    <td>76</td>
    <td>15485</td>
  </tr>
</table>

<p class="persian align-right">
  طبق این جدول می توان نتیجه گرفت:
</p>
<ul class="persian align-right">
  <li>
    همه ی این الگوریتم ها کامل هستند.
  </li>
  <li>
    استفاده از الگوریتم weighted A Star لزوما جواب بهینه را نتیجه نمیدهد.
  </li>
  <li>
    می توان از نظر تعداد استیت های دیده شده تا جواب، این الگوریتم ها را به شکل زیر مقایسه کرد:
    <br/>
    Weighted A Star < A Star < BFS < IDS
  </li>
</ul>

<p class="persian align-right">
  در نهایت برای مقایسه ی این الگوریتم ها به طور کلی می توان گفت که:
</p>
<ul class="persian align-right">
  <li>
    مزیت BFS سادگی پیاده سازی آن است
  </li>
  <li>
    مزیت IDS استفاده ی بهینه از حافظه است چرا که از O(db) حافظه استفاده شده است. ( d طول مسیر بهینه و b مقدار branching factor است )
  </li>
  <li>
    مزیت A Star تعداد استیت های بررسی شده ی کمتر آن است که در بلند مدت باعث کاهش زمان اجرا می شود. اما به هر حال وجود آگاهی در رابطه با محیط، شرط استفاده از این الگوریتم است.
  </li>
  <li>
    مزیت Weighted A Star نیز سرعت بالای آن است. اما این سرعت بالا با بهینه بودن جواب به گونه ای مبادله شده است.
  </li>
</ul>